cmd/client.go

```diff
@@ -244,6 +254,135 @@ type Config struct {
 	Transport         http.RoundTripper
 }
 
+// getCredsChain returns an []credentials.Provider array for the config
+// and the STS configuration (if present)
+func (config *Config) getCredsChain() ([]credentials.Provider, *probe.Error) {
+	var credsChain []credentials.Provider
+	// if an STS endpoint is set, we will add that to the chain
+	if stsEndpoint := env.Get("MC_STS_ENDPOINT_"+config.Alias, ""); stsEndpoint != "" {
+		// set AWS_WEB_IDENTITY_TOKEN_FILE is MC_WEB_IDENTITY_TOKEN_FILE is set
+		if val := env.Get("MC_WEB_IDENTITY_TOKEN_FILE_"+config.Alias, ""); val != "" {
+			os.Setenv("AWS_WEB_IDENTITY_TOKEN_FILE", val)
+			if val := env.Get("MC_ROLE_ARN_"+config.Alias, ""); val != "" {
+				os.Setenv("AWS_ROLE_ARN", val)
+			}
+			if val := env.Get("MC_ROLE_SESSION_NAME_"+config.Alias, randString(32, rand.NewSource(time.Now().UnixNano()), "mc-session-name-")); val != "" {
+				os.Setenv("AWS_ROLE_SESSION_NAME", val)
+			}
+		}
+		stsEndpointURL, err := url.Parse(stsEndpoint)
+		if err != nil {
+			return nil, probe.NewError(fmt.Errorf("Error parsing sts endpoint: %w", err))
+		}
+		credsSts := &credentials.IAM{
+			Client: &http.Client{
+				Transport: config.getTransport(),
+			},
+			Endpoint: stsEndpointURL.String(),
+		}
+		credsChain = append(credsChain, credsSts)
+	}
+
+	signType := credentials.SignatureV4
+	if strings.EqualFold(config.Signature, "s3v2") {
+		signType = credentials.SignatureV2
+	}
+
+	// Credentials
+	creds := &credentials.Static{
+		Value: credentials.Value{
+			AccessKeyID:     config.AccessKey,
+			SecretAccessKey: config.SecretKey,
+			SessionToken:    config.SessionToken,
+			SignerType:      signType,
+		},
+	}
+	credsChain = append(credsChain, creds)
+	return credsChain, nil
+}
+
+// getTransport returns a corresponding *http.Transport for the *Config
+// set withS3v2 bool to true to add traceV2 tracer.
+func (config *Config) getTransport() http.RoundTripper {
+	if config.Transport == nil {
+		config.initTransport(true)
+	}
+	return config.Transport
+}
+
+func (config *Config) isTLS() bool {
+	if stsEndpoint := env.Get("MC_STS_ENDPOINT_"+config.Alias, ""); stsEndpoint != "" {
+		stsEndpointURL, err := url.Parse(stsEndpoint)
+		if err != nil {
+			return false
+		}
+		return isHostTLS(config) || stsEndpointURL.Scheme == "https"
+	}
+	return isHostTLS(config)
+}
+
+func (config *Config) initTransport(withS3v2 bool) {
+	var transport http.RoundTripper
+
+	useTLS := config.isTLS()
```

Comment:
This could be moved to the inner scope (line 349), but it's not an expensive operation, so fine with me to leave it here. As it is used only once, you could also replace the check in line 349 with `config.isTLS()`.